modules/common/react/lib/utils/models.ts

```diff
@@ -5,14 +5,17 @@ export type Model<State, Events> = {
   events: Events;
 };
 
+// bivarianceHack is used for force bivariance of the function constraint. Without this, it will fail with `strictFunctionTypes`
+type IEvent = {[key: string]: {bivarianceHack(data?: object): void}['bivarianceHack']};
```

Comment:
This bivariance hack is the same as what's used in React. This is a case for valid bivariance - I want to set a constraint of an `Events` object. It must be a record of functions that either accept an object or nothing. This is to make callback and guard signatures predictable through the `createEventMap` and `useEventMap` utility functions. `strictFunctionTypes` complain about this bivariance though since a real event will not have a signature of `(data: object) => void` and will look like `(data: { foo: string }) => void` which is a contravariance violation. But it is valid to set up constraints in this case, so a bivariance hack is necessary.

Here's a simplified example of what's going on: [Typescript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYgrgOwMZQLxQN4CMCWA3AQwCccDkIAJApAawAoATA4AgfgC4oB7LAKwiTAAlJzxccDKAF8A2gCJchEmSSVqNOQF0AUAHpdoSLEQp0jZgU49+goWgB8UMRKj6oSAnADOEL+64IwEQESqTkTjhcADbMkQhQAO44wAAWUF5BOILwyMBxACrgvtraAGYmeQHuRBDMEDlIADz5UBAAHsAQCAx+DfZ0pQic+XYY2lBQNcBwRPGD2lIlSAEZ1bWdDA1oa3UNdOYsnBhQpVxcnBkkCADm0naojmMTUzMIC0JLK8BQALYgMNRgFwiCBts1Wh0uj1uHwBMB+jgEGA4MBhvdHuNJhBprMoIjkcAFp8EKtOj8wNs-gDBMCQHQMKdzlAAORYYjMqRCIA)

Reply:
i really appreciate the description here plus the playground!